compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/util/IrUtils.kt

```diff
@@ -286,7 +286,7 @@ fun IrSimpleFunction.collectRealOverrides(): Set<IrSimpleFunction> {
 // This implementation is from kotlin-native
 // TODO: use this implementation instead of any other
 fun IrSimpleFunction.resolveFakeOverride(): IrSimpleFunction? {
-    return collectRealOverrides().singleOrNull { it.modality != Modality.ABSTRACT }
+    return collectRealOverrides().firstOrNull { it.modality != Modality.ABSTRACT }
```

Comment:
I think there is an assumption that we could have only one real implementation for a fake override and it would be nice to keep it. I'd suggest adding another implementation with Deprecated annotation instead of modifying this one.

Reply:
@udalov @bashor Thanks for the review. The original fix I had was more localized (similar to what @bashor is recommending) but I went this route because the comment above `resolveFakeOverride` mentioned it was from K/N, and the implementation for K/N was using `first`: https://github.com/JetBrains/kotlin-native/blob/bf3d5965df8df0999dfd6d3439c407090c4c5de9/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/descriptors/DescriptorUtils.kt#L78 

Aside from `DefaultArgumentStubGenerator.parametersForCall`, there was another problematic usage of `resolveFakeOverride` (that assumes only one real implementation) that I am not sure is fixed by KOTLIN-CR-3792. It's this one in `InterfaceLowering`: https://github.com/JetBrains/kotlin/blob/c9df17f2f18700702db3d844047ac5556edea911/compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt#L96

This is in fact why I had two tests. `compiler/testData/codegen/box/defaultArguments/function/kt36188_2.kt` covers the problematic assumption in `InterfaceLowering`. If this is also fixed in KOTLIN-CR-3792, then I'm ok to just wait for that to merge with these tests added. I didn't go through that CR thoroughly but I _don't_ think it's fixed, in which case I can just make the localized fix: Add a new implementation that uses `firstOrNull` and apply it in the 2 problematic call sites.

Reply:
@punzki You're right, I've missed the usage in `InterfaceLowering`.

In this case it seems reasonable to add a new function which returns `firstOrNull` and use it in these two call sites (I don’t think annotating it with Deprecated, as proposed by @bashor, is correct though). Thanks!

While you're at it, consider also improving the exception message when no fake override is found. Right now it's just NPE because of `!!` at call sites, but we could make the function itself check if no overrides are found and render the function to ease debugging:
```
fun IrSimpleFunction.resolveFakeOverrideToFirst(): IrSimpleFunction {
    return collectRealOverrides().firstOrNull { it.modality != Modality.ABSTRACT }
        ?: error("No override is found: ${render()}")
}
```

Reply:
My point is that for fake overrides we shouldn't have more than one overridden declaration with an implementation modulo bugs. 
But in the case of default arguments, it's slightly different -- we have to find first with default arguments independent to modality, just finding the first implementation is not enough. And in this case, it's "firstOrNull" but not "singleOrNull" only because of the issue in FE.

Probably we have to rename these functions to avoid feature misuses.

Reply:
BTW, current JVM BE generates wrong code for these valid cases:
```kotlin
interface Q {
    fun foo(a: Int) {}
}

interface Z {
    fun foo(a: Int = 1)
}

class R : Q, Z {
    override fun foo(a: Int) {
        println(a)
    }
}

fun main() {
    R().foo()
}
```

```kotlin
interface Q {
    fun foo(a: Int = 1) {}
}

interface Z {
    fun foo(a: Int)
}

class R : Z, Q {
    override fun foo(a: Int) {
        println(a)
    }
}

fun main() {
    R().foo()
}
```

Reply:
I didn't check them yet for IR BEs and didn't find related issues in YT

Reply:
Some related issues
https://youtrack.jetbrains.com/issue/KT-15971
https://youtrack.jetbrains.com/issue/KT-9474
https://youtrack.jetbrains.com/issue/KT-34866

Reply:
@bashor Can you clarify what you mean by "But in the case of default arguments, it's slightly different -- we have to find first with default arguments independent to modality, just finding the first implementation is not enough."?

You mean we should just find the first overridden method, even if it _is_ abstract? Why is that?

Reply:
Sorry, if I confused you.

>  first with default arguments
I meant the first declaration where the default value is defined and it could have any modality (see examples above).
It's about how we should find declaration if we need default values (specifically in `DefaultArgumentStubGenerator#parametersForCall`)

In `InterfaceLowering` we have to find the first `*$default` with implementation.
Another point was that we could be more strict here and use "single" if we didn't have a bug in FE.

Reply:
BTW, examples above look like work in JVM IR BE

Reply:
@bashor Sorry for the delayed reply. In `DefaultArgumentStubGenerator` we are trying to resolve fake overrides for `*$default` and not the original function. All "real" (_not_ fake overrides) `*$default` functions will _never_ have "abstract" modality; they are always "final" since there's always some implementation eventually (the implementation that calls the original function). So while it's okay to find the first "real" function regardless of modality, it should also be okay to keep the `{ it.modality != Modality.ABSTRACT }` filter predicate.

I will also add those examples. Those are KT-15971, right?

Reply:
@bashor @udalov Rebased and made the requested changes, now in `BridgeLowering` instead of `DefaultArgumentStubGenerator`. Please take a look!

I also added your examples as tests. I named the new function `resolveFakeOverrideWithBugCompatibility`; hopefully this makes it very clear that it's there only for bug-compatibility purposes, without having to mark it `@Deprecated`.